src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,976 @@ mod v1 {
+    #[derive(Default)]
+    struct CfFile {
+        pub cf: CfName,
+        pub path: PathBuf,
+        pub tmp_path: PathBuf,
+        pub sst_writer: Option<SstFileWriter>,
+        pub file: Option<File>,
+        pub kv_count: u64,
+        pub size: u64,
+        pub written_size: u64,
+        pub checksum: u32,
+        pub write_digest: Option<Digest>,
+    }
+
+    #[derive(Default)]
+    struct MetaFile {
+        pub meta: SnapshotMeta,
+        pub path: PathBuf,
+        pub file: Option<File>,
+
+        // for writing snapshot
+        pub tmp_path: PathBuf,
+    }
+
+    pub struct Snap {
+        display_path: String,
+        cf_files: Vec<CfFile>,
+        cf_index: usize,
+        meta_file: MetaFile,
+        size_track: Arc<RwLock<u64>>,
+    }
+
+    impl Snap {
+        fn new<T: Into<PathBuf>>(dir: T,
+                                 key: &SnapKey,
+                                 size_track: Arc<RwLock<u64>>,
+                                 sending: bool,
+                                 to_build: bool)
+                                 -> RaftStoreResult<Snap> {
+            let dir_path = dir.into();
+            if !dir_path.exists() {
+                try!(fs::create_dir_all(dir_path.as_path()));
+            }
+            let snap_prefix = if sending {
+                SNAP_GEN_PREFIX
+            } else {
+                SNAP_REV_PREFIX
+            };
+            let prefix = format!("{}_{}", snap_prefix, key);
+            let display_path = Snap::get_display_path(&dir_path, &prefix);
+
+            let mut cf_files = Vec::with_capacity(SNAPSHOT_CFS.len());
+            for cf in SNAPSHOT_CFS {
+                let filename = format!("{}_{}{}", prefix, cf, SST_FILE_SUFFIX);
+                let path = dir_path.join(&filename);
+                let tmp_path = dir_path.join(format!("{}{}", filename, TMP_FILE_SUFFIX));
+                let cf_file = CfFile {
+                    cf: cf,
+                    path: path,
+                    tmp_path: tmp_path,
+                    ..Default::default()
+                };
+                cf_files.push(cf_file);
+            }
+
+            let meta_filename = format!("{}{}", prefix, META_FILE_SUFFIX);
+            let meta_path = dir_path.join(&meta_filename);
+            let meta_tmp_path = dir_path.join(format!("{}{}", meta_filename, TMP_FILE_SUFFIX));
+            let meta_file = MetaFile {
+                path: meta_path,
+                tmp_path: meta_tmp_path,
+                ..Default::default()
+            };
+
+            let mut s = Snap {
+                display_path: display_path,
+                cf_files: cf_files,
+                cf_index: 0,
+                meta_file: meta_file,
+                size_track: size_track,
+            };
+
+            // load snapshot meta if meta_file exists
+            if file_exists(&s.meta_file.path) {
+                if let Err(e) = s.load_snapshot_meta() {
+                    if !to_build {
```

Comment:
The meta is needed in the most common interface `exists()`. `Snap` is usually used as a short-time living object in the context. It's not shared or reused. It would not benefit to delay the meta loading, but just make the code complicated.